Improve the output format of check-webkit-style

rr-codes opened the pull request:
#### d7aca415c23a1b5d182ba4359077378edafe04d4
<pre>
Improve the output format of check-webkit-style
<a href="https://bugs.webkit.org/show_bug.cgi?id=295140">https://bugs.webkit.org/show_bug.cgi?id=295140</a>
<a href="https://rdar.apple.com/154540370">rdar://154540370</a>

Reviewed by Brianna Fan.

Make the output formatted a bit nicer, and also with color. The format is inspired by swift-format.

A screenshot of after vs. before will be attached to the PR.

* Tools/Scripts/webkitpy/common/system/logtesting.py:
(LogTesting.setUp):

Adjust the testing logger to match the updated logic of the normal logging mechanism.

* Tools/Scripts/webkitpy/style/checker.py:
(_create_log_handlers):
(StyleProcessorConfiguration.write_style_error):

Introduce a new logging level &quot;style&quot; (with value 41, which is 1 greater than &quot;error&quot;). This is so
that code style related errors are formatted in a unique way; specifically, the level is not logged
at the beginning of the string.

Consequently, a new handler is introduced to differentiate between style errors and other errors.

* Tools/Scripts/webkitpy/style/checker_unittest.py:
(StyleProcessorConfigurationTest.test_write_style_error_emacs):
(StyleProcessorConfigurationTest.test_write_style_error_vs7):
(StyleProcessor_EndToEndTest.test_process):
(StyleProcessor_CodeCoverageTest.test_invalid_file):
* Tools/Scripts/webkitpy/style/error_handlers_unittest.py:
(DefaultStyleErrorHandlerTest.test_max_reports_per_category):
(DefaultStyleErrorHandlerTest.test_line_numbers):
* Tools/Scripts/webkitpy/style/main_unittest.py:

Adjust test expectations for the new format.

Canonical link: <a href="https://commits.webkit.org/296935@main">https://commits.webkit.org/296935@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1f6d2247956635299ea3f75531944033461a5fb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110007 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29666 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20098 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116029 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60255 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38253 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83634 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112955 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24193 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99064 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64076 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/109439 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23565 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17212 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59824 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93566 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17255 "Found 1 new test failure: imported/w3c/web-platform-tests/feature-policy/feature-policy-frame-policy-timing.https.sub.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118820 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37047 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27463 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92607 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37419 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95330 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92431 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23555 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37423 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15151 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32929 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36941 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36603 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39943 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38312 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -669,7 +669,7 @@ def _create_log_handlers(stream):
     error_handler = logging.StreamHandler(stream)
     error_handler.setLevel(logging.WARNING)
-    formatter = logging.Formatter("%(levelname)s: %(message)s")
+    formatter = logging.Formatter("%(message)s")
```
You got rid of a format specifier here. This is what's causing most of the test fails

```
[1845/2718] webkitpy.style.error_handlers_unittest.DefaultStyleErrorHandlerTest.test_max_reports_per_category erred:
  Traceback (most recent call last):
    File "/Volumes/Data/worker/WebKitPy-Tests-EWS/build/Tools/Scripts/webkitpy/style/error_handlers_unittest.py", line 143, in test_max_reports_per_category
      self._call_error_handler(error_handler, confidence)
    File "/Volumes/Data/worker/WebKitPy-Tests-EWS/build/Tools/Scripts/webkitpy/style/error_handlers_unittest.py", line 75, in _call_error_handler
      handle_error(line_number=line_number,
    File "/Volumes/Data/worker/WebKitPy-Tests-EWS/build/Tools/Scripts/webkitpy/style/error_handlers.py", line 160, in __call__
      self._configuration.write_style_error(category=category,
    File "/Volumes/Data/worker/WebKitPy-Tests-EWS/build/Tools/Scripts/webkitpy/style/checker.py", line 1049, in write_style_error
      _log.error(format_string % (file_path,
  TypeError: not enough arguments for format string
```

rr-codes replied:
the problem isn't that line, it's

```
        if self._output_format == 'vs7':
            format_string = "%s(%s):  %s  [%s] [%d]"
        else:
            format_string = "\x1b[1m%s:%s:\x1b[0m \x1b[31;1merror:\x1b[0m \x1b[33;1m[%s]\x1b[0m \x1b[1m%s\x1b[0m"


        _log.error(format_string % (file_path,
                                    line_number,
                                    category,
                                    message))
```

in vs7 mode

aprotyas commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -1043,13 +1043,13 @@ def write_style_error(self,
             format_string = "%s(%s):  %s  [%s] [%d]"
         else:
-            format_string = "%s:%s:  %s  [%s] [%d]"
+            format_string = "\x1b[1m%s:%s:\x1b[0m \x1b[31;1merror:\x1b[0m \x1b[33;1m[%s]\x1b[0m \x1b[1m%s\x1b[0m"
```
You might want to name these escape codes so they're easier to read? i.e.

```py
bright = "\x1b[1m"
reset = "\x1b[0m"

```

briannafan replied:
Agreed! We could also use this as a chance to introduce f-strings (although it might be worse to have a Frankensteinian file of various format types...)

aprotyas reviewed (commented):
Uh-oh 😅 Do you mind adding a few more words about what you expect with the changes, maybe a before/after in the commit log? The colors in t

> Found 10 webkitpy test failures: webkitpy.style.checker_unittest.ConfigureLoggingTest.test_warning_message, webkitpy.style.checker_unittest.StyleProcessorConfigurationTest.test_write_style_error_emacs, webkitpy.style.main_unittest.ExpectationLinterInStyleCheckerTest.test_linter_added_file_with_error, webkitpy.style.main_unittest.ExpectationLinterInStyleCheckerTest.test_linter_deleted_file, webkitpy.style.main_unittest.ExpectationLinterInStyleCheckerTest.test_linter_duplicate_line, webkitpy.style.checker_unittest.StyleProcessorConfigurationTest.test_write_style_error_vs7, webkitpy.style.checker_unittest.StyleProcessor_CodeCoverageTest.test_invalid_file, webkitpy.style.checker_unittest.StyleProcessor_EndToEndTest.test_process, webkitpy.style.error_handlers_unittest.DefaultStyleErrorHandlerTest.test_line_numbers, webkitpy.style.error_handlers_unittest.DefaultStyleErrorHandlerTest.test_max_reports_per_category

briannafan commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -1043,13 +1043,13 @@ def write_style_error(self,
-                                           line_number,
-                                           message,
-                                           category,
-                                           confidence_in_error))
```
It might be worth asking around if having the confidence displayed is useful to anyone before removing it!

rr-codes replied:
not only does it seem useless, it seems the vast majority didn't even know what it was!

briannafan commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -1041,15 +1053,19 @@ def write_style_error(self,
+            reset = "\x1b[0m"
+            white = "\x1b[1m"
+            red = "\x1b[31;1m"
+            yellow = "\x1b[33;1m"
```
Could we set these as global variables, then import them into the unit tests as needed to prevent duplication? And on the off-chance these get changed it would be easier to manage.

briannafan commented on Tools/Scripts/webkitpy/common/system/logtesting.py:
```diff
@@ -147,19 +147,33 @@ def setUp(test_case, logging_level=logging.INFO):
+        logging.addLevelName(41, "STYLE")
+
         stream = TestLogStream(test_case)
         handler = logging.StreamHandler(stream)
```
Can we call this `style_handler` to stay consistent with what's in `checker.py`?

briannafan commented on Tools/Scripts/webkitpy/common/system/logtesting.py:
```diff
@@ -147,19 +147,33 @@ def setUp(test_case, logging_level=logging.INFO):
         handler = logging.StreamHandler(stream)
-        handler.setLevel(logging_level)
-        formatter = logging.Formatter("%(levelname)s: %(message)s")
+        handler.setLevel(41)
```
Where you hardcode 41, can we call `logging.getLevelName("STYLE")`? Same goes for the usage in `checker.py`

briannafan commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -666,9 +666,21 @@ def _create_log_handlers(stream):
+    # The filter method accepts a logging.LogRecord instance.
+    error_filter.filter = lambda record: record.levelno < 41
+
+    # Handles logging.ERROR.
```
This comment made me confused... Can we clarify it's also handling `WARNING` logs?

briannafan commented on Tools/Scripts/webkitpy/style/checker.py:
```diff
@@ -666,9 +666,21 @@ def _create_log_handlers(stream):
-    # Handles logging.WARNING and above.
+    logging.addLevelName(41, "STYLE")
+
+    style_handler = logging.StreamHandler(stream)
```
Let's add a comment similar to the ones below like `Handles logging.STYLE`

briannafan commented on Tools/Scripts/webkitpy/style/main_unittest.py:
```diff
@@ -203,9 +206,16 @@ def test_linter_added_file_with_error(self):
+        red = "\x1b[31;1m"
+        yellow = "\x1b[33;1m"
+
+        expected = f"{white}/mock-checkout/LayoutTests/TestExpectations:3:{reset} {red}error:{reset} {yellow}[test/expectations]{reset} {white}Duplicate or ambiguous entry lines LayoutTests/TestExpectations:2 and LayoutTests/TestExpectations:3.{reset}\n"
```
nit: Let's use `ANSIColor` here as well!

briannafan commented on Tools/Scripts/webkitpy/style/main_unittest.py:
```diff
@@ -220,9 +230,16 @@ def test_linter_deleted_file(self):
+        red = "\x1b[31;1m"
+        yellow = "\x1b[33;1m"
+
+        expected = f"{white}/mock-checkout/LayoutTests/TestExpectations:2:{reset} {red}error:{reset} {yellow}[test/expectations]{reset} {white}Path does not exist.{reset}\n"
```
nit: More `ANSIColor` :)

briannafan reviewed (approved):
r+ with two minor changes to use `ANSIColor`

rr-codes committed:
Improve the output format of check-webkit-style
https://bugs.webkit.org/show_bug.cgi?id=295140
rdar://154540370

Reviewed by Brianna Fan.

Make the output formatted a bit nicer, and also with color. The format is inspired by swift-format.

A screenshot of after vs. before will be attached to the PR.

* Tools/Scripts/webkitpy/common/system/logtesting.py:
(LogTesting.setUp):

Adjust the testing logger to match the updated logic of the normal logging mechanism.

* Tools/Scripts/webkitpy/style/checker.py:
(_create_log_handlers):
(StyleProcessorConfiguration.write_style_error):

Introduce a new logging level "style" (with value 41, which is 1 greater than "error"). This is so
that code style related errors are formatted in a unique way; specifically, the level is not logged
at the beginning of the string.

Consequently, a new handler is introduced to differentiate between style errors and other errors.

* Tools/Scripts/webkitpy/style/checker_unittest.py:
(StyleProcessorConfigurationTest.test_write_style_error_emacs):
(StyleProcessorConfigurationTest.test_write_style_error_vs7):
(StyleProcessor_EndToEndTest.test_process):
(StyleProcessor_CodeCoverageTest.test_invalid_file):
* Tools/Scripts/webkitpy/style/error_handlers_unittest.py:
(DefaultStyleErrorHandlerTest.test_max_reports_per_category):
(DefaultStyleErrorHandlerTest.test_line_numbers):
* Tools/Scripts/webkitpy/style/main_unittest.py:

Adjust test expectations for the new format.

Canonical link: https://commits.webkit.org/296935@main
